runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java

```diff
@@ -0,0 +1,199 @@
+      try {
+        List<? extends Source<T>> partitionedSources = source.splitIntoBundles(desiredSizeBytes,
+            runtimeContext.getPipelineOptions());
+        Partition[] partitions = new SourcePartition[partitionedSources.size()];
+        for (int i = 0; i < partitionedSources.size(); i++) {
+          partitions[i] = new SourcePartition<>(id(), i, partitionedSources.get(i));
+        }
+        return partitions;
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to create partitions.", e);
+      }
+    }
+
+    @Override
+    public scala.collection.Iterator<WindowedValue<T>> compute(final Partition split,
+                                                               TaskContext context) {
+      final Iterator<WindowedValue<T>> iter = new Iterator<WindowedValue<T>>() {
+        @SuppressWarnings("unchecked")
+        SourcePartition<T> partition = (SourcePartition<T>) split;
+        BoundedSource.BoundedReader<T> reader = createReader(partition);
+
+        private boolean finished = false;
+        private boolean started = false;
+        private boolean closed = false;
+
+        @Override
+        public boolean hasNext() {
+          try {
+            if (!started) {
+              started = true;
+              finished = !reader.start();
+            } else {
+              finished = !reader.advance();
+            }
+            if (finished) {
+              // safely close the reader if there are no more elements left to read.
+              closeIfNotClosed();
+            }
+            return !finished;
+          } catch (IOException e) {
+            throw new RuntimeException("Failed to read from reader.", e);
+          }
+        }
+
+        @Override
+        public WindowedValue<T> next() {
+          return WindowedValue.of(reader.getCurrent(), reader.getCurrentTimestamp(),
+              GlobalWindow.INSTANCE, PaneInfo.NO_FIRING);
+        }
+
+        @Override
+        public void remove() {
+          // do nothing.
+        }
+
+        private void closeIfNotClosed() {
+          if (!closed) {
+            closed = true;
+            try {
+              reader.close();
+            } catch (IOException e) {
+              throw new RuntimeException("Failed to close Reader.", e);
+            }
+          }
+        }
+      };
+
+      return new InterruptibleIterator<>(context,
+          scala.collection.JavaConversions.asScalaIterator(iter));
+    }
+
+    private BoundedSource.BoundedReader<T> createReader(SourcePartition<T> partition) {
+      try {
+        return ((BoundedSource<T>) partition.source).createReader(
+            runtimeContext.getPipelineOptions());
+      } catch (IOException e) {
+        throw new RuntimeException("Failed to create reader from a BoundedSource.", e);
+      }
+    }
+  }
+
+  /**
+   * An input {@link Partition} wrapping the partitioned {@link Source}.
+   */
+  private static class SourcePartition<T> implements Partition {
```

Comment:
`AutoValue` seems like an overkill here, though I will change `hashCode()` a little bit to be more like Spark's other implementations like: https://github.com/apache/spark/blob/branch-1.6/core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala#L47